From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: OverwriteMC <artyom030803@gmail.com>
Date: Thu, 2 Apr 2026 20:10:28 +0300
Subject: [PATCH] Cache identifier toString and hash

Cache toString and hash to reduce allocations.
Use lazy cache here to prevent cache inconsistency caused by
operations that don't use the constructor, e.g., Gson serialization or Reflection.

diff --git a/net/minecraft/resources/Identifier.java b/net/minecraft/resources/Identifier.java
index df28b83ab34b2e44a6ce2dbb903ef94db5496aaa..d7807975ae2a608bdf2937d68e3c0f71a5fd3cc2 100644
--- a/net/minecraft/resources/Identifier.java
+++ b/net/minecraft/resources/Identifier.java
@@ -26,6 +26,10 @@ public final class Identifier implements Comparable<Identifier> {
     public static final String PAPER_NAMESPACE = "paper"; // Paper
     private final String namespace;
     private final String path;
+    // Leaf start - Cache identifier toString and hash
+    private String identifierStr;
+    private int hash;
+    // Leaf end - Cache identifier toString and hash
 
     private Identifier(final String namespace, final String path) {
         assert isValidNamespace(namespace);
@@ -131,7 +135,14 @@ public final class Identifier implements Comparable<Identifier> {
 
     @Override
     public String toString() {
-        return this.namespace + ":" + this.path;
+        // Leaf start - Cache identifier toString and hash
+        String identifierStr = this.identifierStr;
+        if (identifierStr == null) {
+            identifierStr = this.namespace + ":" + this.path;
+            this.identifierStr = identifierStr;
+        }
+        return identifierStr;
+        // Leaf end - Cache identifier toString and hash
     }
 
     @Override
@@ -141,7 +152,14 @@ public final class Identifier implements Comparable<Identifier> {
 
     @Override
     public int hashCode() {
-        return 31 * this.namespace.hashCode() + this.path.hashCode();
+        // Leaf start - Cache identifier toString and hash
+        int hash = this.hash;
+        if (hash == 0) {
+            hash = 31 * this.namespace.hashCode() + this.path.hashCode();
+            this.hash = hash;
+        }
+        return hash;
+        // Leaf end - Cache identifier toString and hash
     }
 
     @Override
